### ISSUE TITLE
Font fallback support

### ISSUE THREAD

**Issue by whereswaldon** [go-text/shaping#8]
Our current `Input` type essentially encodes the expectation that the runes provided with it can be displayed using the single font face provided. This pushes detecting whether a font face can display runes onto a hypothetical higher-level API that would split text along the boundaries of font support and create `Input`s for each block that can be handled by a font.

However, harfbuzz does provide a way for us to detect when it couldn't find glyphs for runes in the input. I think it probably makes more sense to change `Input` to have `Faces []font.Face` that are tried in order, and to change

```go
type Shaper interface {
    Shape(Input) Output
}
```
into
```go
type Shaper interface {
    Shape(Input) []Output
}
```
So that the shaper can break the input into multiple outputs each in one of the provided faces. This would require `Output` to have a field indicating the font face in use.

Does this seem like a reasonable way to do it? I basically think we need font fallback, and the current way seems really awkward. I'd love other ideas for how to handle it.

**Comment by andydotxyz** [comment]
This is a powerful suggestion, though how does it fit with things like line mixed LTR/RTL?
In previous discussions we decided it was up to the consumer of this API to pre-split these items.
What you propose implies to me that we could move that back into scope for `shaping`?

**Comment by andydotxyz** [comment]
> Does this seem like a reasonable way to do it? I basically think we need font fallback, and the current way seems really awkward. I'd love other ideas for how to handle it.

It seems like a good idea. In Fyne we do something similar, look to see if a font supports a glyph before asking for it to be shaped / drawn. We have the "user requested font" falling back to theme font then finally a builtin font.

**Comment by whereswaldon** [comment]
> This is a powerful suggestion, though how does it fit with things like line mixed LTR/RTL?
In previous discussions we decided it was up to the consumer of this API to pre-split these items.
What you propose implies to me that we could move that back into scope for shaping?

Well, I guess I see this as a different problem. The question "does font X support a glyph for rune Y" is actually quite complicated due to ligatures and substitution rules. The easiest way to answer it seems to be to shape the text and check whether the font couldn't resolve some glyphs adequately. Given that, it seemed natural to me to handle that for the user.

For instance, let's say that we're trying to shape this string: `Hello 안녕하십니까`. Let's assume this is user provided, so we don't know in advance that it contains multiple languages. This string should be presented in a single style, and is all a single direction, so it isn't obvious that we need to split it into multiple `Input` unless we perform a font support test on every rune.

With the current API, we'd need to perform the glyph support test (again, not sure how to do this reliably any faster than shaping the text) to determine where to break this into two inputs. We'd provide different fonts inside of each `Input`, and harfbuzz would do its thing.

With the proposed API, we provide a prioritized slice of fonts in the `Input`, and we immediately shape the text. Our shaper discovers the substring for which the first font fails to provide glyphs, and it tries the subsequent fonts until one that supports Korean fills in the gaps. It returns `[]Output`, one with each font used.

I'd also be happy to leave this out of the shaper, but wrap the shaping process with a "font resolving" process that did this. It definitely can happen at a higher level.

**Comment by andydotxyz** [comment]
I agree that this is a different problem, and the reasons for doing it are compelling.
However from a developers point of view it seems like a similar level challenge as font splitting into RTL/LTR chunks. There are technical reasons that they are implemented differently, but does that mean we should expose them as different? Does it make sense from the developers point of view, and/or does it fit the scope of the repo to fallback based on missing glyphs but not to change direction based on the glyph content?

**Comment by whereswaldon** [comment]
I wouldn't be opposed to splitting bidi text into directional runs within this repo. You're right that it appears similar, though I believe that splitting bidi is easier to do as a precursor step since it requires no font data (just the unicode codepoints). My current line wrapper has a comment like `//TODO: split bidi here`, so I definitely think it's in scope. The bidi splitting would convert a single `Input` into `[]Input`, whereas the font fallback does something similar to the output.

As far as code organization, since this task is somewhat linked to shaping (cannot perform font fallback without doing the shaping), I think it makes sense to surface it here. Do you have somewhere else that you'd rather see the feature exposed?

**Comment by andydotxyz** [comment]
It had not occurred to me that a codepoint always used the same direction.
My reading had me think it was more nuanced than that.

As for code organisation I did not have an alternative to suggest, I was just asking the question. From what you have said I guess it makes sense to expand the scope here.

I think there is one more complication though. Fonts can specify a fallback character... so how do we interpret that? override it? ignore it? try all fonts and fallback to first font's fallback, or the last?

**Comment by npillmayer** [comment]
> It had not occurred to me that a codepoint always used the same direction. My reading had me think it was more nuanced than that.

You're right there: it is more complicated. Just as a quick note: There are "strong directional characters", which imply a writing direction. But there are weak and neutral character categories as well. If you zoom out enough, the Unicode Bidi Algorithm is about determining the "zones of contagion" of strong chars to the others, until LTR and RTL zones touch.

On the topic of splitting into runs: Bidi runs and attribute-runs do not form a hierarchy (in general). Attribute-runs (except color) imply a change in font, but for bidi-runs it's less clear.

Cheers,
–Norbert

**Comment by whereswaldon** [comment]
> It had not occurred to me that a codepoint always used the same direction.
> My reading had me think it was more nuanced than that.

Sorry, I didn't mean to imply that it was solely-dependent on the codepoint. It's dependent on all of the codepoints together. You run a unicode BiDi Algorithm on all of the text as @npillmayer described to segment it. The important characteristic here is that you don't need any font data to do so. It's all based on the codepoints themselves.

**Comment by andydotxyz** [comment]
Hmm, good points.
Does this impact our inputs? Should we consider `[]codepoint`, with another library going from `string` to `[]codepoint`? Or is it best to leave as string?
Just a random thought - alternatively we could pass the original string along with line demarcation and RTL/LTR hints (a separate slice of markers?).
Am thinking how can we minimise extra memory use or unnecessary processing.

> cannot perform font fallback without doing the shaping

I have thought more about this. Isn't it true that a shaping from a font will always point to glyphs that the font contains? If so then fallback could be done before or after I think.
Thinking out lout, isn't fallback therefore better in part of the rendering process, that is where we take glyph and font info to create what is rendered?

**Comment by whereswaldon** [comment]
>I have thought more about this. Isn't it true that a shaping from a font will always point to glyphs that the font contains? If so then fallback could be done before or after I think.

This was my first intuitive thought, but I've seen some counterexamples. Symbol fonts, for example, may provide a ellipsis glyph like `…` that is used when someone types the string `etc`. The font matches that rune _sequence_ against the glyph using a dynamic rule. The font may not provide glyphs for the characters `e`, `t`, and `c`. Only shaping the string will reveal that the font could, in fact, resolve the `etc` sequence to the glyph `…`.

Cases like the above are why I think the shaper has to be involved. Iterating the runes and checking for support in the font will only catch simple cases, and may incorrectly conclude that a font lacks support for a rune sequence.

> Does this impact our inputs? Should we consider []codepoint, with another library going from string to []codepoint? Or is it best to leave as string?

The go `rune` type is the same thing as `codepoint`. There is an argument for using unicode grapheme clusters instead of runes for some tasks, but I think we should worry about that later. That's primarily a concern for text editing, not for display.

Our current input type allows you to specify a subrange of characters that are actually relevant to the current shaping. This means that it can be used for bidi by creating multiple Inputs with a shared []rune, and different indices of content. Example:

Consider the string `Sarah (שרה)`. We apply an as-yet-unwritten bidi algorithm to create three inputs:
- The first one (`Sarah (`) has `Start: 0, End: 8` and direction LTR
- The second one (`שרה`) has `Start: 8, End: 14` and direction RTL
- The third one (`)`) has `Start: 14, End: 15` and direction LTR

They can all share the same `[]rune`, so we're not copying the string data unnecessarily.

**Comment by andydotxyz** [comment]
> The font may not provide glyphs for the characters e, t, and c. Only shaping the string will reveal that the font could, in fact, resolve the etc sequence to the glyph ….

I don't get how this is a fallback - the font recognises the codepoints and knows what glyph to use. What I was meaning is that the glyph must exist in the current font for that rule to apply - so fallback would happen after this transition, potentially in a separate process?

Unless I have misunderstood what fallback is intended to be?

> Our current input type allows you to specify a subrange of characters that are actually relevant to the current shaping.

Thanks, it had not occurred to me that the same `rune[]` would be passed in multiple `Input` types. And you're quite right, `rune` is fine in place of the codepoint discussion.

**Comment by whereswaldon** [comment]
Let's say that we have two fonts. Font1 is a symbol font providing many fancy symbols (including the etc=>… rewrite discussed above), but no letter glyphs. Font2 is a normal latin font.

We want to shape the string `lions, tigers, bears, etc`.

We try the fonts in order using the naive approach of checking if each rune in the string is present in the font. When checking each rune, we conclude that none of the runes in the string have glyphs provided by Font1, as it only provides symbols. It does not provide `e`, `t`, or `c`. We then check the next font in our list of fonts and shape the whole string in Font2. The resulting string is `lions, tigers, bears, etc` in Font2.

If we instead use the shaper, we shape the whole string in Font1 and get a glyph sequence that looks like: `□□□□□□ □□□□□□□ □□□□□□ …` (assuming that the symbol font provides a space glyph). We see that there are three sub-sequences composed of the "couldn't find a glyph" glyph, and we retry shaping those using Font2. This means we arrive at the shaped result `lions, tigers, bears, …`. A probable optimization is to merge subsequences separated by whitespace so we only reshape once instead of three times.

This same kind of fallback would be important if you had a separate font providing special ligatures that you wanted to try before a normal font. Since ligatures are generally multiple runes fusing together into a single glyph, a simple "is there a glyph for this rune" lookup will not detect them.

**Comment by andydotxyz** [comment]
I may be wrong - but it was my understanding that such ligatures have their own codepoint? The examples you illustrate seem to be a transformation that go from one (or a set of) codepoints to another.
Is this really shaping? It sounds like pre-processing.

As you indicate this could multiply out processing time and shaping is probably one of the slowest parts of text handling, so if we can pre-process instead this could be faster.

**Comment by whereswaldon** [comment]
> I may be wrong - but it was my understanding that such ligatures have their own codepoint?

It is true that [unicode has codepoints for ligatures](https://codepoints.net/U+FB02). However, I still expect to see a ligature when I type `fl` in Roboto. I typed two separate characters, resulting in two separate runes. The _font_ has a rule that will map that rune sequence to a ligature, but without shaping I can't know whether the font will do that or not.

>  The examples you illustrate seem to be a transformation that go from one (or a set of) codepoints to another.

Sorry, I think I was unclear! It's mapping from sequences of codepoints to _glyphs_, not other codepoints. The text's representation in binary is unchanged, but the font has a lot of freedom in how it presents that data visually. Harfbuzz allows us to leverage a lot of the cool (and complex) features of OpenType fonts, but at the cost of dealing with complexities like these.

It's tricky to talk about this here, because I have to craft sequences of unicode codepoints that _look like_ the glyphs the shaper would emit. I can't directly control the font rendering of my github markdown, so I can't _actually_ demonstrate what I mean. So my example text does literally map from codepoints to codepoints, but my meaning is that the second one is what the glyphs emitted by the shaper _look like_.

**Comment by npillmayer** [comment]
> > I may be wrong - but it was my understanding that such ligatures have their own codepoint?
> 
> It is true that [unicode has codepoints for ligatures](https://codepoints.net/U+FB02). However, I still expect to see a ligature when I type `fl` in Roboto. I typed two separate characters, resulting in two separate runes. The _font_ has a rule that will map that rune sequence to a ligature, but without shaping I can't know whether the font will do that or not.

In general, fonts may (and certainly do for complex scripts) have lots of ligature glyphs without a mapping from a code-point. If clients do not start with the code-point sequence to produce the ligature, these ligatures are literally not to be found by the client.

> but without shaping I can't know whether the font will do that or not

That captures it nicely :-)  The CMap table in a font is a one-way street.

> I can't directly control the font rendering of my github markdown, so I can't actually demonstrate what I mean

http://npillmayer.github.io/UAX/img/CP-2-GID.png

Things get even more interesting in the case of glyph *reordering*, which is a *common case* for complex scripts :-D

**Comment by benoitkugler** [comment]
My understanding of the Pango way of doing things is actually to only rely on input information to select the font to use, and not to use Harfbuzz result. 
I think the reason is simply that it would be too costly to try and shape the same input with many fonts... 

More precisely, what Pango does is :  

- split the given input text to item with constant font description. "Font description" is a bit vague, but it includes for instance family, weight, italic) 
- for one "font description", query all the fonts available matching this description. This is done on Linux by fontconfig. In general it is not an easy task, but it could be simplified to just using the user provided fonts.
- among this set of fonts, choose the first one covering the input rune. It is _actually_ a rune, and not a glyph, meaning that indeed some substitution may be missed
- split again the input to group the runes using the same font, and finally use Harbuzz on each group


To sum up, Pango actually only use the CMap information of the fonts, and not all the possible rune transformations supported by Harfbuzz, to select the font. 
I have no information on the possible limitation of this approach...

**Comment by whereswaldon** [comment]
@benoitkugler Interesting! I trust that the pango authors know what they're doing. This seems like a tradeoff in performance that you might choose differently based on requirements. I'm really curious to see benchmarks for how much slower one option is than the other.

Given that, I'd be happy for us to provide a function that tests for rune-by-rune support in a font and segments an input in that way. This is both faster and simpler, and probably covers the GUI use-case (though it can miss some substitutions). I'm still interested in exploring the other option, but it's better to get the easy thing working first.

**Comment by npillmayer** [comment]
> To sum up, Pango actually only use the CMap information of the fonts, and not all the possible rune transformations supported by Harfbuzz, to select the font. I have no information on the possible limitation of this approach...

> @benoitkugler Interesting! I trust that the pango authors know what they're doing. This seems like a tradeoff in performance that you might choose differently based on requirements. I'm really curious to see benchmarks for how much slower one option is than the other.

@whereswaldon, I'm not sure I understand where this is heading. Font selection can *only* be done by code-point coverage + language tag; there is no other possible way**. I'm not sure what the other option is supposed to be. Your "etc" example is making things too complicated, I'm afraid – or not complicated enough, depending on your point of view:

* It is irrelevant whether a code-point produces a visible glyph in a font, as long as it has a CMap entry and therefore produces a glyph index. This glyph index may be used as input in GSUB to trigger rules to form your "etc" ligature. But it is unpredictable from the outside which glyph position the ligature will be at and where the ligature rule finally will point at.
* For "etc" with font 1, it *has to have* code-points in CMap for 'e', 't' and 'c', regardless of wether it has sensible glyph outlines stored for them at that position. Otherwise the GSUB table could not get information on what to combine in the first place. Moreover, each of 'e', 't' and 'c' on their own will produce a glyph with font 1, even if it would be an invisible zero-width spacer, but no `.notdef` (glyph index 0).
* It is unpredictable and more or less untracable (except for the vague information of HarfBuzz's cluster-ID), what the shaper and font 1 decide to produce the glyph output for "etc". It may be a single glyph, or a combination of 17 repositioned glyphs in a super-fancy way.
* There is (almost) no such thing as "HarfBuzz rune transformations". Transformations are generally done on glyphs. The shaper supports the OpenType engine by supplying strategically placed feature requests before triggering the OT Layout rules, which operate solely on glyph indices. (There are exceptions: Transformations on the *rune*/*code-point* level actually exist for things like composing/de-composing of accented letters and the like – heavily used in complex scripts, cases where Harfbuzz compensates for common errors in input sequences, cases for scripts where POS knowledge is necessary for correct shaping, etc.)
* Code-point to glyph selection may be locale-specific in a font. The most complex case of font coverage decision is East Asian Han script variations. In general, this case requires additional knowledge about a font covering Han/CJK which locale it prefers/is mainly design for.

**) Except of course user markup, style-sheets of the text stack, pre-selected or default fonts, etc.

**Comment by khaledhosny** [comment]
There are two fallback strategies:
- check font coverage before shaping and itemize text based on characters supported by the main font and subsequent fonts in the font fallback list. By far this is the most commonly implemented strategy, but it has several drawbacks because for example HarfBuzz might synthetize characters not supported by the font (e.g. various spaces), or handle control characters without needing a glyph in the font at all (e.g. ZWJ, ZWNJ), or apply Unicode normalization based on what the font sypports (e.g. if the input text uses NFC or NFD form but the font supports the other form). Implementations using this strategy might apply similar tricks themselves, but this is seldom done.
- shape the text with harfbuzz then for runs of characters HarfBuzz decides the font does not support, shape with next font in the fallback list and so on untill done. This avoids the drawbacks of the first strategy, but can be costly. This is the stretegy HarfBuzz developers advocate. Chrome implments this and employs several tricks to avoid too much reshaping. LibreOffice also does this but only for the main font, for the fallback fonts it does simple font coverage check first.

**Comment by npillmayer** [comment]
> There are two fallback strategies:

Thanks for the clarification, @khaledhosny! Yes, these two options exists. However, *both* would have solved @whereswaldon's "etc" example. That's because this premise does not hold:

> When checking each rune, we conclude that none of the runes in the string have glyphs provided by Font1

The "naive approach" mentioned is effectively your option 1. 

I'm not advocating checking coverage by the app, nor to rely on the shaper. Both have their challenges from an app level perspective. From the discussion above I'm just seeing some misconceptions about what an app may be able to anticipate when looking at a font. From an app perspective, it's best to think about shaping as

    code-point input   |----(Magic happens)---->    glyph output

**Comment by whereswaldon** [comment]
Thanks everyone for providing your expertise. I'm certainly new at this, and it's clear that my mental model isn't totally correct. I'm going to trust @npillmayer about the invalidity of my `etc` example. It does makes sense that the font would have to have some kind of data for those characters in order to trigger the rule that substitutes a fancy glyph.

I'll rephrase my [earlier post](https://github.com/go-text/shaping/issues/8#issuecomment-984955387):

I think we should implement a helper function that checks font coverage before shaping and automatically breaks an input into runs that can be shaped in a given font. This has plenty of drawbacks that @khaledhosny and @npillmayer have pointed out, but it's the fastest and simplest option available.

**Later**, I think we should also offer a function that will use the more expensive re-shaping strategy instead. It may well be that this is fast enough for the GUI toolkits if we optimize sufficiently, but it seems like significantly more time/effort to build.

**Comment by benoitkugler** [comment]
What do you think of the following signature for the helper (naming aside) : 

```
// SplitByCoverage split the runes from 'input' to several items, with the same 
// characteristics as 'input', but with the first font among 'availableFonts' providing support for all the rune
// of the item.
// Coverage is obtained by calling the NominalGlyph() method of each font.
SplitByCoverage(input Input, availableFonts []font.Face) []Input
```

I'm not sure where the language is useful. I don't think it is used in CMaps. (It is used to activate/deactivate some glyphs substitutions/positioning, but we have opt to not use Harfbuzz logic inside this helper.)

**Comment by whereswaldon** [comment]
That looks great to me!

**Comment by npillmayer** [comment]
> I'm not sure where the language is useful. I don't think it is used in CMaps. (It is used to activate/deactivate some glyphs substitutions/positioning, but we have opt to not use Harfbuzz logic inside this helper.)

Strange as it may sound, the language *is* part of the keys for CMap. From the OpenType spec: "The encoding record entries in the 'cmap' header must be sorted first by platform ID, then by platform-specific encoding ID, and then by the **language field** in the corresponding subtable." The language field is used for the Macintosh-platform only.

Additionally, as I wrote above, there is the complication of Han unification:

> Code-point to glyph selection may be locale-specific in a font. The most complex case of font coverage decision is East Asian Han script variations. In general, this case requires additional knowledge about a font covering Han/CJK which locale it prefers/is mainly design for.

Apple may not be good at choosing API function names, but the signature looks attractive, IMHO: [MacOS CoreText Font Fallback](https://developer.apple.com/documentation/coretext/1509992-ctfontcopydefaultcascadelistforl). Its output would be fed into your `SplitByCoverage(…)`.

**Comment by andydotxyz** [comment]
When picking language I think we should keep in mind our audience.
As it's not aimed at font experts it might be more in-keeping to pick names that describe in lay-mans terms instead of correct terminologies.
Perhaps `SplitByFontGlyphs` or something. It would probably mirror (if we have a separate method for) bi-di splits, which could be `SplitByDirectionGlyphs` (just thinking out loud).

**Comment by benoitkugler** [comment]
> Strange as it may sound, the language _is_ part of the keys for CMap. From the OpenType spec: "The encoding record entries in the 'cmap' header must be sorted first by platform ID, then by platform-specific encoding ID, and then by the **language field** in the corresponding subtable." The language field is used for the Macintosh-platform only.

Thanks for the precision. I've checked the  [Harfbuzz code](https://github.com/harfbuzz/harfbuzz/blob/main/src/hb-ot-cmap-table.hh) and the language field of the cmap is **actually ignored**, despite being indeed in the spec.
As a starter, I would suggest to stay close to Harfbuzz and ignore the language as well.

**Comment by benoitkugler** [comment]
With https://github.com/go-text/font/pull/3 merged, I can propose a PR, unless somebody is already working on it ?

**Comment by whereswaldon** [comment]
Nope, I'm not working on this right now. :D My next contribution will be line wrapping, once I'm completely sure it works for the use-cases I need it for.

**Comment by andydotxyz** [comment]
Go ahead. I am currently looking at the abstraction side of things (which I need to re-work on the new repository once it lands).
I'd like to keep the API that we expose to go-text only and with the new repo it should be even easier to maintain that :)
